crypto-bigint: add modulo operations to `UInt`

tarcieri opened the pull request:
Continuing on @dignifiedquire's work from #510, this PR attempts to simplify the implementation.

~~Aspirationally, the goal is to add equivalence tests with a known-good implementation.~~ Proptests against `num-bigint` added.

dignifiedquire commented on crypto-bigint/src/uint/add_mod.rs:
```diff
@@ -1,123 +1,97 @@
-            }
+use crate::{Limb, UInt};
+
+impl<const LIMBS: usize> UInt<LIMBS> {
```
Given that we do not have specialization yet, this now prevents me from adding optimizations for specific limb sizes, which are critical for optimal performance

tarcieri replied:
What if we have a generic `const fn` version that always works, and traits which are specific to limb sizes?

That's how all of the existing macros in this crate work anyway... they write trait impls.

dignifiedquire replied:
fair, that should work as well

tarcieri commented:
Added some trait wrappers in 2b37f48. A couple notes:

- used macros to impl them for `1..=12` limbs, but you can potentially omit some of those and add different impls
- they have the same name as the `const fn` which is a bit annoying/confusing, but has mostly worked out in the elliptic curve crates
- I added `debug_assert`s for `< p`

Ideally we can eventually replace that with `const impl` and specialization in some hypothetical future.

tarcieri commented:
@dignifiedquire I tried to write a proptest for `add_mod` in 54a3c47, but it's failing:

https://github.com/RustCrypto/utils/pull/561/checks?check_run_id=3196567313

However the code looks correct to me, so I'm not sure if there's something wrong with the test itself or the code.
